src/util/heap/layout/vm_layout_constants.rs

```diff
@@ -120,48 +34,130 @@ pub const LOG_PAGES_IN_SPACE64: usize = LOG_SPACE_SIZE_64 - LOG_BYTES_IN_PAGE as
-#[cfg(target_pointer_width = "64")]
-pub const SPACE_MASK_64: usize = ((1 << LOG_MAX_SPACES) - 1) << SPACE_SHIFT_64;
+impl VMLayoutConstants {
+    /// Normal 32-bit configuration
+    pub const fn new_32bit() -> Self {
+        Self {
+            log_address_space: 32,
+            heap_start: chunk_align_down(unsafe { Address::from_usize(0x8000_0000) }),
+            heap_end: chunk_align_up(unsafe { Address::from_usize(0xd000_0000) }),
+            vm_space_size: chunk_align_up(unsafe { Address::from_usize(0xdc0_0000) }).as_usize(),
+            log_max_chunks: Self::LOG_ARCH_ADDRESS_SPACE - LOG_BYTES_IN_CHUNK,
+            log_space_extent: 31,
+            space_shift_64: 0,
+            space_mask_64: 0,
+            space_size_64: 1 << 31,
+            force_use_contiguous_spaces: false,
+        }
+    }
+    /// Normal 64-bit configuration
+    #[cfg(target_pointer_width = "32")]
+    pub fn new_64bit() -> Self {
+        unimplemented!("64-bit heap constants do not work with 32-bit builds")
+    }
+    #[cfg(target_pointer_width = "64")]
+    pub fn new_64bit() -> Self {
+        Self {
+            log_address_space: 47,
+            heap_start: chunk_align_down(unsafe {
+                Address::from_usize(0x0000_0200_0000_0000usize)
+            }),
+            heap_end: chunk_align_up(unsafe { Address::from_usize(0x0000_2200_0000_0000usize) }),
+            vm_space_size: chunk_align_up(unsafe { Address::from_usize(0xdc0_0000) }).as_usize(),
+            log_max_chunks: Self::LOG_ARCH_ADDRESS_SPACE - LOG_BYTES_IN_CHUNK,
+            log_space_extent: 41,
+            space_shift_64: 41,
+            space_mask_64: ((1 << 4) - 1) << 41,
+            space_size_64: 1 << 41,
+            force_use_contiguous_spaces: true,
+        }
+    }
 
-/*
- * Size of each space in the 64-bit memory layout
- *
- * We can't express this constant in a 32-bit environment, hence the
- * conditional definition.
- */
-// FIXME: When Compiling for 32 bits this expression makes no sense
-// #[allow(const_err)]
-// pub const SPACE_SIZE_64: usize = if_then_else_usize!(HEAP_LAYOUT_64BIT,
-//    1 << LOG_SPACE_SIZE_64, MAX_SPACE_EXTENT);
-#[cfg(target_pointer_width = "64")]
-pub const SPACE_SIZE_64: usize = 1 << LOG_SPACE_SIZE_64;
-#[cfg(target_pointer_width = "32")]
-pub const SPACE_SIZE_64: usize = MAX_SPACE_EXTENT;
+    /// Custom VM layout constants. VM bindings may use this function for compressed or 39-bit heap support.
+    /// This function must be called before MMTk::new()
+    pub fn set_custom_vm_layout_constants(constants: VMLayoutConstants) {
+        let mut guard = CUSTOM_CONSTANTS.lock();
+        assert!(
+            guard.is_none(),
+            "Custom VM_LAYOUT_CONSTANTS can only be set once"
+        );
+        *guard = Some(constants);
+    }
+}
+
+static CUSTOM_CONSTANTS: Mutex<Option<VMLayoutConstants>> = Mutex::new(None);
+
+lazy_static! {
+    pub static ref VM_LAYOUT_CONSTANTS: VMLayoutConstants = {
```

Comment:
If it is determined at boot time, it is no longer constants.  So it always requires a memory load no matter if it is loaded from a global static struct or a field of MMTk.  I prefer calling them properties of a MMTk instance, and they should be stored in read-only fields in the MMTk instance, and propagated to sub-components if needed.

An alternative is making them part of the `VMBinding` trait so that they can really be constants.  If `MMTK<VM>` is parameterised by `<VM>`, the Rust compiler knows the constants in `VM` at compile time when specialising `MMTK` for the `VM`.  Even if those properties are expressed as methods in `VMBinding`, if the concrete implementation of those functions return constant values, the compiler can still inline those functions at compile time.

Reply:
> If it is determined at boot time, it is no longer constants. So it always requires a memory load no matter if it is loaded from a global static struct or a field of MMTk. I prefer calling them properties of a MMTk instance, and they should be stored in read-only fields in the MMTk instance, and propagated to sub-components if needed.

The VM map depends on the VM layout. But our VM map is currently global. I don't think we can or we should make the layout as a field in an MMTk instance without changing the VM map. I would think we just leave the VM layout in the same way as the VM map, and we can refactor both in the future if we need to.

> An alternative is making them part of the `VMBinding` trait so that they can really be constants. If `MMTK<VM>` is parameterised by `<VM>`, the Rust compiler knows the constants in `VM` at compile time when specialising `MMTK` for the `VM`. Even if those properties are expressed as methods in `VMBinding`, if the concrete implementation of those functions return constant values, the compiler can still inline those functions at compile time.

Those are not constants. For OpenJDK, it is determined by a command line option `-XX:+UseCompressedOops`. I think we would see the performance results, then decide if runtime VM layout should be a feature or not. If the feature is enabled, we get the layout from the binding and store it as a variable (as it is now). If the feature is not enabled, the VM layout is a constant (either defined in MMTk core, or defined in the `VMBinding`).

Reply:
Hm. I think that making this into a feature might be unnecessary. I would say it's pretty standard to want different starting heap addresses, address space sizes, etc.

Reply:
> Hm. I think that making this into a feature might be unnecessary. I would say it's pretty standard to want different starting heap addresses, address space sizes, etc.

Yeah, we should allow bindings to set those info. The proposed feature is about whether we allow to set them at run-time (as opposed to compile-time). We should know if runtime VM layout introduces any overhead. If it introduces some overhead, it is reasonable to make it into a feature so it won't slow down bindings that do not need run-time VM layout.

Reply:
Right. It would be good to test the performance of this change.

Reply:
Just finished the benchmark. Looks like there is no noticeable slowdown

Immix: http://squirrel.anu.edu.au/plotty-public/wenyuz/v8/p/NkK9YG
GenImmix: http://squirrel.anu.edu.au/plotty-public/wenyuz/v8/p/StQwU3

SS and GenSS are also included in the run, but they didn't finish all the benchmarks due to the small heap factor.

Reply:
Just checking that this is with or without PGO?

Reply:
without pgo

Reply:
The performance results look good.